File: sdk/go/hashring_test.go
package nanocached

import (
	"errors"
	"fmt"
	"math"
	"reflect"
	"sort"
	"testing"
)

func TestFnv1aMatchesPublishedVectors(t *testing.T) {
	cases := map[string]uint64{
		"":       0xcbf29ce484222325,
		"a":      0xaf63dc4c8601ec8c,
		"foobar": 0x85944171f73967e8,
	}
	for input, want := range cases {
		if got := fnv1a([]byte(input)); got != want {
			t.Errorf("fnv1a(%q) = %#x, want %#x", input, got, want)
		}
	}
}

func TestMatchesCrossLanguageScoreVectors(t *testing.T) {
	// Pinned outputs of the full client-side replication score pipeline — the Rust node
	// and the TypeScript/Python/Java/Rust/.NET SDKs assert these too.
	if got := fmix64(0); got != 0 {
		t.Errorf("fmix64(0) = %#x", got)
	}
	if got := fmix64(1); got != 0xb456bcfc34c2cb2c {
		t.Errorf("fmix64(1) = %#x", got)
	}
	if got := fmix64(0xcbf29ce484222325); got != 0xefd01f60ba992926 {
		t.Errorf("fmix64(offset basis) = %#x", got)
	}

	ring := NewHashRing([]string{"node-a", "node-b", "node-c"})
	cases := map[string][]string{
		"alpha": {"node-c", "node-b", "node-a"},
		"beta":  {"node-a", "node-c", "node-b"},
		"":      {"node-a", "node-b", "node-c"},
	}
	for key, want := range cases {
		if got := ring.Owners([]byte(key), 3); !reflect.DeepEqual(got, want) {
			t.Errorf("Owners(%q) = %v, want %v", key, got, want)
		}
	}
}

// ── namespaces (issue #105) ──────────────────────────────────────────

// TestNamespacedKeyHashMatchesCrossLanguageVectors pins keyHash's
// namespaced form (fnv1a(be32(len(ns)) || ns || key)) against the same
// vectors the server (src/hash_ring.rs) and the other five SDKs assert,
// next to the existing alpha/beta/"" vectors above.
func TestNamespacedKeyHashMatchesCrossLanguageVectors(t *testing.T) {
	cases := []struct {
		ns, key string
		want    uint64
	}{
		{"users", "alpha", 0xfd4ab55027c21df6},
		{"users", "", 0xa9e9bbca44bb502e}, // hash-only vector; the wire itself rejects an empty key
		{"\xff\x00", "beta", 0x8f7c097eccb8e792},
	}
	for _, c := range cases {
		if got := keyHash([]byte(c.ns), []byte(c.key)); got != c.want {
			t.Errorf("keyHash(%q, %q) = %#x, want %#x", c.ns, c.key, got, c.want)
		}
	}
}

// TestNamespacedOwnersMatchCrossLanguageVectors is the same vectors run
// through the full HRW pipeline, over the ring/replicas the issue #105
// spec pins them against.
func TestNamespacedOwnersMatchCrossLanguageVectors(t *testing.T) {
	ring := NewHashRing([]string{"node-a", "node-b", "node-c"})
	cases := []struct {
		ns, key string
		want    []string
	}{
		{"users", "alpha", []string{"node-a", "node-c", "node-b"}},
		{"users", "", []string{"node-b", "node-c", "node-a"}},
		{"\xff\x00", "beta", []string{"node-b", "node-a", "node-c"}},
	}
	for _, c := range cases {
		got := ring.OwnersNS([]byte(c.ns), []byte(c.key), 3)
		want := c.want
		if !reflect.DeepEqual(got, want) {
			t.Errorf("OwnersNS(%q, %q) = %v, want %v", c.ns, c.key, got, want)
		}
	}
}

// TestDefaultNamespaceHashesAndRoutesExactlyLikeTheLegacyForm is the
// rolling-upgrade invariant from hash_ring.rs's module docs: an
// unnamespaced key's placement must not move when namespaces enter the
// picture — nil and "" must both behave as "no namespace at all", and
// OwnersNS(nil, ...)/Owners(...) must agree byte-for-byte with the
// pre-#105 alpha vector above.
func TestDefaultNamespaceHashesAndRoutesExactlyLikeTheLegacyForm(t *testing.T) {
	if got, want := keyHash(nil, []byte("alpha")), fnv1a([]byte("alpha")); got != want {
		t.Errorf("keyHash(nil, \"alpha\") = %#x, want %#x", got, want)
	}
	if got, want := keyHash([]byte(""), []byte("alpha")), fnv1a([]byte("alpha")); got != want {
		t.Errorf("keyHash(\"\", \"alpha\") = %#x, want %#x", got, want)
	}

	ring := NewHashRing([]string{"node-a", "node-b", "node-c"})
	want := []string{"node-c", "node-b", "node-a"}
	if got := ring.OwnersNS(nil, []byte("alpha"), 3); !reflect.DeepEqual(got, want) {
		t.Errorf("OwnersNS(nil, \"alpha\") = %v, want %v", got, want)
	}
	if got := ring.OwnersNS([]byte(""), []byte("alpha"), 3); !reflect.DeepEqual(got, want) {
		t.Errorf("OwnersNS(\"\", \"alpha\") = %v, want %v", got, want)
	}
	if got := ring.Owners([]byte("alpha"), 3); !reflect.DeepEqual(got, want) {
		t.Errorf("Owners(\"alpha\") = %v, want %v", got, want)
	}
}

// TestNamespaceAndKeyBoundariesAreUnambiguous: length-prefixing the
// namespace is what keeps ("ab","c") and ("a","bc") from colliding, and
// keeps a namespaced key from ever colliding with the un-namespaced
// concatenation of the two — mirrors hash_ring.rs's own boundary test.
func TestNamespaceAndKeyBoundariesAreUnambiguous(t *testing.T) {
	if got, other := keyHash([]byte("ab"), []byte("c")), keyHash([]byte("a"), []byte("bc")); got == other {
		t.Errorf("keyHash(\"ab\",\"c\") == keyHash(\"a\",\"bc\") == %#x", got)
	}
	if got, legacy := keyHash([]byte("ab"), []byte("c")), keyHash(nil, []byte("abc")); got == legacy {
		t.Errorf("keyHash(\"ab\",\"c\") == keyHash(nil,\"abc\") == %#x", got)
	}
}

func TestOwnersAreDistinctAndCapped(t *testing.T) {
	ring := NewHashRing([]string{"node-a", "node-b", "node-c"})
	owners := ring.Owners([]byte("some-key"), 2)
	if len(owners) != 2 || owners[0] == owners[1] {
		t.Errorf("Owners = %v", owners)
	}
	if got := ring.Owners([]byte("some-key"), 10); len(got) != 3 {
		t.Errorf("capped Owners = %v", got)
	}
}

// TestOwnersMatchesANaiveFullSortReference pins Owners' output
// byte-identical to the straightforward "sort everything, then
// truncate" reference it replaced (the bounded top-R selection it now
// does instead must never change which nodes come back, or their
// order) across the edge cases in replicas (0, 1, exactly the node
// count, and over it) plus a run of pseudo-random keys.
func TestOwnersMatchesANaiveFullSortReference(t *testing.T) {
	type scored struct {
		score uint64
		node  string
	}
	naiveOwners := func(ring *HashRing, key []byte, replicas int) []string {
		keyHash := fnv1a(key)
		ranked := make([]scored, len(ring.nodes))
		for i, node := range ring.nodes {
			ranked[i] = scored{fmix64(ring.nodeHashes[i] ^ keyHash), node}
		}
		sort.Slice(ranked, func(a, b int) bool {
			if ranked[a].score != ranked[b].score {
				return ranked[a].score > ranked[b].score
			}
			return ranked[a].node < ranked[b].node
		})
		if replicas > len(ranked) {
			replicas = len(ranked)
		}
		if replicas <= 0 {
			return []string{}
		}
		owners := make([]string, replicas)
		for i := range owners {
			owners[i] = ranked[i].node
		}
		return owners
	}

	nodes := make([]string, 37)
	for i := range nodes {
		nodes[i] = fmt.Sprintf("node-%d", i)
	}
	ring := NewHashRing(nodes)

	// A tiny deterministic PRNG (splitmix64) stands in for a "math/rand"
	// seeded sequence, keeping this test free of any flakiness tied to
	// the runtime's default source.
	state := uint64(0xC0FFEE)
	next := func() uint64 {
		state += 0x9E3779B97F4A7C15
		z := state
		z = (z ^ (z >> 30)) * 0xBF58476D1CE4E5B9
		z = (z ^ (z >> 27)) * 0x94D049BB133111EB
		return z ^ (z >> 31)
	}

	for _, replicas := range []int{0, 1, len(nodes) / 2, len(nodes), len(nodes) + 1, len(nodes) + 10} {
		for i := 0; i < 200; i++ {
			key := []byte(fmt.Sprintf("fuzz-key-%d", next()))
			got := ring.Owners(key, replicas)
			want := naiveOwners(ring, key, replicas)
			if !reflect.DeepEqual(got, want) {
				t.Fatalf("replicas=%d key=%q: Owners = %v, want %v", replicas, key, got, want)
			}
		}
	}
}

func TestAddingANodeNeverReordersExistingNodes(t *testing.T) {
	before := NewHashRing([]string{"node-a", "node-b", "node-c"})
	after := NewHashRing([]string{"node-a", "node-b", "node-c", "node-d"})
	for i := 0; i < 500; i++ {
		key := []byte(fmt.Sprintf("key-%d", i))
		var newOrder []string
		for _, node := range after.Owners(key, 4) {
			if node != "node-d" {
				newOrder = append(newOrder, node)
			}
		}
		if !reflect.DeepEqual(before.Owners(key, 3), newOrder) {
			t.Fatalf("reordered for key-%d", i)
		}
	}
}

func TestRouteErrorsOnAnEmptyRing(t *testing.T) {
	node, err := NewHashRing(nil).Route([]byte("k"))
	if !errors.Is(err, ErrInvalidArgument) {
		t.Fatalf("err = %v, want ErrInvalidArgument", err)
	}
	if node != "" {
		t.Fatalf("node = %q, want empty", node)
	}
}

func TestSpreadsKeysEvenly(t *testing.T) {
	nodes := []string{"node-a", "node-b", "node-c"}
	ring := NewHashRing(nodes)
	counts := map[string]int{}
	const total = 3000
	for i := 0; i < total; i++ {
		node, err := ring.Route([]byte(fmt.Sprintf("key-%d", i)))
		if err != nil {
			t.Fatal(err)
		}
		counts[node]++
	}
	fair := float64(total) / float64(len(nodes))
	for node, count := range counts {
		if math.Abs(float64(count)-fair)/fair >= 0.15 {
			t.Errorf("%s got %d/%d", node, count, total)
		}
	}
}

File: sdk/java/src/test/java/org/nanocached/HashRingTest.java
package org.nanocached;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.Test;

class HashRingTest {
    private static final List<String> NODES = List.of("node-a", "node-b", "node-c");

    private static byte[] bytes(String text) {
        return text.getBytes(StandardCharsets.UTF_8);
    }

    @Test
    void matchesPublishedFnv1aVectors() {
        assertEquals(0xcbf29ce484222325L, HashRing.fnv1a(bytes("")));
        assertEquals(0xaf63dc4c8601ec8cL, HashRing.fnv1a(bytes("a")));
        assertEquals(0x85944171f73967e8L, HashRing.fnv1a(bytes("foobar")));
    }

    @Test
    void matchesCrossLanguageScoreVectors() {
        // Pinned outputs of the full client-side replication score pipeline — the Rust,
        // TypeScript, and Python implementations assert these too.
        assertEquals(0L, HashRing.fmix64(0));
        assertEquals(0xb456bcfc34c2cb2cL, HashRing.fmix64(1));
        assertEquals(0xefd01f60ba992926L, HashRing.fmix64(0xcbf29ce484222325L));

        HashRing ring = new HashRing(NODES);
        assertEquals(List.of("node-c", "node-b", "node-a"), ring.owners(bytes("alpha"), 3));
        assertEquals(List.of("node-a", "node-c", "node-b"), ring.owners(bytes("beta"), 3));
        assertEquals(List.of("node-a", "node-b", "node-c"), ring.owners(bytes(""), 3));
    }

    // Namespaces (issue #105): `key_hash = fnv1a(be32(len(ns)) || ns || key)`
    // for a non-empty namespace. Every SDK and the server pin these same
    // owner-order vectors (computed independently in Python from the
    // spec's definition) over the identical node-a/node-b/node-c ring, so
    // a disagreement here means this SDK's routing would silently diverge
    // from the rest of the cluster.
    @Test
    void matchesCrossLanguageNamespacedVectors() {
        HashRing ring = new HashRing(NODES);
        assertEquals(List.of("node-a", "node-c", "node-b"), ring.owners(bytes("users"), bytes("alpha"), 3));
        assertEquals(List.of("node-b", "node-c", "node-a"), ring.owners(bytes("users"), bytes(""), 3));
        assertEquals(List.of("node-b", "node-a", "node-c"),
                ring.owners(new byte[] {(byte) 0xff, 0x00}, bytes("beta"), 3));
    }

    @Test
    void theDefaultNamespaceHashesExactlyLikeTheLegacyForm() {
        // Rolling-upgrade invariant: an un-namespaced key's placement must
        // not move when the server (and this SDK) learns about namespaces.
        HashRing ring = new HashRing(NODES);
        assertEquals(ring.owners(bytes("alpha"), 3), ring.owners(bytes(""), bytes("alpha"), 3));
        assertEquals(ring.owners(bytes(""), 3), ring.owners(bytes(""), bytes(""), 3));
    }

    @Test
    void namespaceAndKeyBoundariesAreUnambiguous() {
        // A delimiter-free split: the length prefix keeps ("ab","c") and
        // ("a","bc") apart, and a namespaced key never collides with the
        // un-namespaced concatenation.
        HashRing ring = new HashRing(NODES);
        assertNotEquals(
                ring.owners(bytes("ab"), bytes("c"), 3),
                ring.owners(bytes("a"), bytes("bc"), 3));
        assertNotEquals(
                ring.owners(bytes("ab"), bytes("c"), 3),
                ring.owners(bytes(""), bytes("abc"), 3));
    }

    @Test
    void namespacesSpreadASharedSingletonKeyOverDifferentNodes() {
        // The reason the namespace is part of the hash input at all: two
        // namespaces sharing a common singleton key name (e.g. "config")
        // must not always land on the same node.
        List<String> nodes = new ArrayList<>();
        for (char c = 'a'; c <= 'h'; c++) nodes.add("node-" + c);
        HashRing ring = new HashRing(nodes);
        Map<String, Integer> primaries = new HashMap<>();
        for (int i = 0; i < 64; i++) {
            String primary = ring.owners(bytes("cache-" + i), bytes("config"), 1).get(0);
            primaries.merge(primary, 1, Integer::sum);
        }
        assertTrue(primaries.size() > 1);
    }

    @Test
    void ownersAreDistinctAndCapped() {
        HashRing ring = new HashRing(NODES);
        List<String> owners = ring.owners(bytes("some-key"), 2);
        assertEquals(2, owners.size());
        assertNotEquals(owners.get(0), owners.get(1));
        assertEquals(3, ring.owners(bytes("some-key"), 10).size());
    }

    @Test
    void addingANodeNeverReordersExistingNodes() {
        HashRing before = new HashRing(NODES);
        HashRing after = new HashRing(List.of("node-a", "node-b", "node-c", "node-d"));
        for (int i = 0; i < 500; i++) {
            byte[] key = bytes("key-" + i);
            List<String> newOrder = after.owners(key, 4).stream()
                    .filter(node -> !node.equals("node-d"))
                    .toList();
            assertEquals(before.owners(key, 3), newOrder, "reordered for key-" + i);
        }
    }

    // Pins owners()' bounded-heap top-R selection byte-identical to the
    // straightforward "score every node, sort descending, truncate"
    // reference it replaced (see HashRing.owners' doc comment) — the
    // bounded selection must never change which nodes come back, or
    // their order, versus a full sort using the identical comparator
    // (descending unsigned score; ties toward the lexicographically
    // smaller name). Mirrors the Go SDK's
    // TestOwnersMatchesANaiveFullSortReference.
    private static List<String> naiveOwners(HashRing ring, List<String> nodes, byte[] key, int replicas) {
        long keyHash = HashRing.fnv1a(key);
        record Scored(long score, String node) {}
        Comparator<Scored> order = (a, b) -> {
            int byScore = Long.compareUnsigned(b.score(), a.score());
            return byScore != 0 ? byScore : a.node().compareTo(b.node());
        };
        List<Scored> ranked = new ArrayList<>();
        for (String node : nodes) {
            ranked.add(new Scored(HashRing.fmix64(HashRing.fnv1a(bytes(node)) ^ keyHash), node));
        }
        ranked.sort(order);
        int limit = Math.min(replicas, ranked.size());
        if (limit <= 0) return List.of();
        List<String> result = new ArrayList<>(limit);
        for (int i = 0; i < limit; i++) result.add(ranked.get(i).node());
        return result;
    }

    @Test
    void ownersMatchesANaiveFullSortReference() {
        List<String> nodes = new ArrayList<>();
        for (int i = 0; i < 37; i++) nodes.add("node-" + i);
        HashRing ring = new HashRing(nodes);

        // A tiny deterministic PRNG (splitmix64) rather than java.util.Random,
        // keeping this test free of any flakiness tied to a seeded
        // default source (mirrors the Go SDK's reference test).
        long[] state = {0xC0FFEEL};
        java.util.function.LongSupplier next = () -> {
            state[0] += 0x9E3779B97F4A7C15L;
            long z = state[0];
            z = (z ^ (z >>> 30)) * 0xBF58476D1CE4E5B9L;
            z = (z ^ (z >>> 27)) * 0x94D049BB133111EBL;
            return z ^ (z >>> 31);
        };

        for (int replicas : new int[] {0, 1, nodes.size() / 2, nodes.size(), nodes.size() + 1, nodes.size() + 10}) {
            for (int i = 0; i < 200; i++) {
                byte[] key = bytes("fuzz-key-" + next.getAsLong());
                assertEquals(naiveOwners(ring, nodes, key, replicas), ring.owners(key, replicas),
                        "replicas=" + replicas + " key=" + new String(key, StandardCharsets.UTF_8));
            }
        }
    }

    // Regression for issue #461 (mirrors src/hash_ring.rs's HashRing::new
    // dedupe from issue #328, and the Python/Go SDK fixes for the same,
    // issues #360/#389): before this fix, a name repeated in the
    // constructor's node list scored independently for each of its
    // slots, so owners() could return the same node name more than once
    // in the top-`replicas` set — inflating its effective share of the
    // ring.
    @Test
    void duplicateNodeNamesAreDeduplicated() {
        HashRing ring = new HashRing(List.of("a", "b", "b", "b", "c", "d"));
        byte[] key = bytes("some-key");

        for (int replicas = 0; replicas <= 5; replicas++) {
            List<String> owners = ring.owners(key, replicas);
            assertEquals(owners.size(), new java.util.HashSet<>(owners).size(),
                    "replicas=" + replicas + " owners=" + owners);
        }

        // Construction order is otherwise unaffected: the first
        // occurrence of a repeated name is kept in place.
        HashRing deduped = new HashRing(List.of("a", "b", "c", "d"));
        for (int i = 0; i < 200; i++) {
            byte[] k = bytes("key-" + i);
            assertEquals(deduped.owners(k, 4), ring.owners(k, 4), "key-" + i);
        }
    }

    @Test
    void spreadsKeysEvenly() {
        HashRing ring = new HashRing(NODES);
        Map<String, Integer> counts = new HashMap<>();
        int total = 3000;
        for (int i = 0; i < total; i++) {
            counts.merge(ring.route(bytes("key-" + i)), 1, Integer::sum);
        }
        double fair = (double) total / NODES.size();
        for (Map.Entry<String, Integer> entry : counts.entrySet()) {
            assertTrue(Math.abs(entry.getValue() - fair) / fair < 0.15,
                    entry.getKey() + " got " + entry.getValue() + "/" + total);
        }
    }
}
